pkg/storage/replica_command.go

```diff
@@ -1304,12 +1329,30 @@ func evalRangeLookup(
 	}
 
 	if len(reply.Ranges) == 0 {
-		// No matching results were returned from the scan. This should
-		// never happen with the above logic.
+		// No matching results were returned from the scan. This can happen when
+		// meta2 ranges split (so for now such splitting is disabled). Remember
+		// that the range addressing keys are generated from the end key of a range
+		// descriptor, not the start key. Consider the scenario:
+		//
+		//   range 1 [a, e):
+		//     b -> [a, b)
+		//     c -> [b, c)
+		//     d -> [c, d)
+		//   range 2 [e, g):
+		//     e -> [d, e)
+		//     f -> [e, f)
+		//     g -> [f, g)
+		//
+		// Now consider looking up the range containing key `d`. The DistSender
+		// routing logic would send the RangeLookup request to range 1 since `d`
+		// lies within the bounds of that range. But notice that the range
+		// descriptor containing `d` lies in range 2. Boom! A real fix will involve
+		// additional logic in the RangeDescriptorCache range lookup state machine.
+		//
+		// See #16266.
```

Comment:
I was planning to close #2266 as the primary issue there is being fixed (we won't access keys during `RangeLookup` that are outside are range) and leave #16266 for properly handling meta2 splits. Agreed that this doesn't really matter.